### PR TITLE
add option to use DST structs for flex arrays

### PR DESCRIPTION
**Pull request by jsgf**
Add `--flexarray-dst` option to allow use of a dynamically sized struct to model a C flexible array member. For example, given the C struct:
```
struct msg {
    unsigned int tag;
    unsigned int len;
    char payload[];
};
```
generate the corresponding Rust type:
```rust
#[repr(C)]
#[derive(Debug)]
pub struct msg<FAM: ?Sized = [::std::os::raw::c_char; 0]> {
    pub tag: ::std::os::raw::c_uint,
    pub len: ::std::os::raw::c_uint,
    pub payload: FAM,
}
```
This single definition is used for both Sized (with a zero-length array as the FAM type) and DST (with a slice).

This also generates the helper methods to convert a raw pointer into a Rust view of the C type, with a given length.
```rust
// Sized implementations
impl msg<[::std::os::raw::c_char; 0]> {
    // Construct a DST reference from a Sized object.
    // SAFETY: there must be at least `len` initialized elements in the underlying storage
    pub unsafe fn flex_ref(
        &self,
        len: usize,
    ) -> &msg<[::std::os::raw::c_char]> {
        unsafe { Self::flex_ptr(self, len) }
    }

    // Same, but mutable
    pub unsafe fn flex_mut_ref(
        &mut self,
        len: usize,
    ) -> &mut msg<[::std::os::raw::c_char]> {
        unsafe { Self::flex_ptr_mut(self, len).assume_init() }
    }

    // Raw pointer variants
    pub unsafe fn flex_ptr<'unbounded>(
        ptr: *const Self,
        len: usize,
    ) -> &'unbounded msg<[::std::os::raw::c_char]> {
        unsafe { &*::std::ptr::from_raw_parts(ptr as *const (), len) }
    }

    pub unsafe fn flex_ptr_mut<'unbounded>(
        ptr: *mut Self,
        len: usize,
    ) -> ::std::mem::MaybeUninit<&'unbounded mut msg<[::std::os::raw::c_char]>>
    {
        unsafe {
            let mut uninit = ::std::mem::MaybeUninit::<
                &mut msg<[::std::os::raw::c_char]>,
            >::uninit();
            (uninit.as_mut_ptr() as *mut *mut msg<[::std::os::raw::c_char]>)
                .write(::std::ptr::from_raw_parts_mut(ptr as *mut (), len));
            uninit
        }
    }
}

// DST implementations
impl msg<[::std::os::raw::c_char]> {
    // Memory size & alignment for allocation
    pub fn layout(len: usize) -> ::std::alloc::Layout {
        unsafe {
            let p: *const Self =
                ::std::ptr::from_raw_parts(::std::ptr::null(), len);
            ::std::alloc::Layout::for_value_raw(p)
        }
    }
    // return the Sized variant along with the length
    pub fn fixed(&self) -> (&msg<[::std::os::raw::c_char; 0]>, usize) {
        unsafe {
            let (ptr, len) = (self as *const Self).to_raw_parts();
            (&*(ptr as *const msg<[::std::os::raw::c_char; 0]>), len)
        }
    }
    pub fn fixed_mut(
        &mut self,
    ) -> (&mut msg<[::std::os::raw::c_char; 0]>, usize) {
        unsafe {
            let (ptr, len) = (self as *mut Self).to_raw_parts();
            (&mut *(ptr as *mut msg<[::std::os::raw::c_char; 0]>), len)
        }
    }
}
```

Upside:
- The flexible array member is directly expressed in a fairly conventional Rust type, with no need to interact with the bindgen-specific `__IncompleteArrayField` type
  - In particular, normal `size_of` will work, taking into account the dynamically sized extension

Problems/TODO:
- depends on unstable `ptr_metadata` and `layout_for_ptr` features
- `from_ptr(_mut)` return references with unbounded lifetimes - it would be up to the caller to constrain them to the underlying storage's lifetime
  - I experimented with a variant which also takes a "lifetime witness" parameter `_lt: &'a LT` which can be used to bound the returned lifetime, but I'll need to try it in practice to see if it works.

This is a prototype for #2771

### CODE REVIEW

**Comment by jsgf** [comment]
Not sure why, but when I run `cargo fmt` locally it wants to change things all over the place.

**Comment by jsgf** [comment]
This doesn't handle the case of a C struct pointing to a DST object (ie thin vs fat pointers).

**Comment by jsgf** [comment]
I haven't checked out the test failures yet - will do so shortly.

I've been using this API for one of my own projects, and so far it feels pretty comfortable.

File: bindgen/codegen/mod.rs [hunk 1]
```diff
@@ -2470,17 +2527,130 @@ impl CodeGenerator for CompInfo {
...
             result.push(quote! {
-                impl #generics Clone for #ty_for_impl {
+                impl #impl_generics_labels Clone for #ty_for_impl {
                     fn clone(&self) -> Self { *self }
                 }
             });
         }
 
+        if needs_flexarray_impl {
+            let prefix = ctx.trait_prefix();
```
Comment (emilio) [inline bindgen/codegen/mod.rs:2542]: Can we move this out to its own function or such? I feel like that'd make it easier to reason about.

File: bindgen-tests/tests/expectations/tests/template-param-usage-9.rs [hunk 2]
```diff
@@ -9,10 +9,10 @@ pub type DoesNotUse_Typedefed<U> = U;
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
 pub struct DoesNotUse_IndirectUsage<T, U> {
-    pub member: DoesNotUse_Aliased<T>,
-    pub another: DoesNotUse_Typedefed<U>,
     pub _phantom_0: ::std::marker::PhantomData<::std::cell::UnsafeCell<T>>,
     pub _phantom_1: ::std::marker::PhantomData<::std::cell::UnsafeCell<U>>,
```
Comment (emilio) [inline bindgen-tests/tests/expectations/tests/template-param-usage-9.rs:13]: Can you send the "move phantom fields to the beginning" as a separate PR? That's fine on its own, and would make this PR a lot easier to review (GitHub's UI for this is not my favourite).

Thanks!

Comment (jsgf) [reply bindgen-tests/tests/expectations/tests/template-param-usage-9.rs:13]: #2783

File: bindgen/ir/comp.rs [hunk 3]
```diff
@@ -825,6 +825,24 @@ impl CompFields {
...
+    /// Return the flex array member for the struct/class, if any.
+    fn flex_array_member(&self, ctx: &BindgenContext) -> Option<TypeId> {
+        let fields = match self {
+            CompFields::Before(_) => panic!("raw fields"),
+            CompFields::After { fields, .. } => fields,
+            CompFields::Error => return None, // panic?
+        };
+
+        // XXX correct with padding on end?
+        match fields.last() {
```
[the remark this answers is not in the document]
Comment (jsgf) [reply bindgen/ir/comp.rs:838]: I'm not sure what you mean.

File: bindgen/codegen/mod.rs (hunk 1 above)
Comment (jsgf) [reply bindgen/codegen/mod.rs:2542]: Yeah, happy to do that. But the rest of the function is so enormous - do you think it should all be factored down?

**Comment by jsgf** [comment]
OK updated with comments (and still on top of #2783)

File: bindgen/ir/comp.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (emilio) [reply bindgen/ir/comp.rs:837]: Sorry, I meant that something like this:

```diff
diff --git a/bindgen/ir/comp.rs b/bindgen/ir/comp.rs
index 31c058ea..f6c9e629 100644
--- a/bindgen/ir/comp.rs
+++ b/bindgen/ir/comp.rs
@@ -834,9 +834,9 @@ impl CompFields {
             CompFields::Error => return None, // panic?
         };

-        match fields.last() {
-            None | Some(Field::Bitfields(..)) => None,
-            Some(Field::DataMember(FieldData { ty, .. })) => ctx
+        match fields.last()? {
+            Field::Bitfields(..) => None,
+            Field::DataMember(FieldData { ty, .. }) => ctx
                 .resolve_type(*ty)
                 .is_incomplete_array(ctx)
                 .map(|item| item.expect_type_id(ctx)),
```

Should compile and work the same.

**Review by emilio — COMMENTED** [review]
@pvdrz do you recall if we have a way of annotating a feature flag as experimental? This is mostly usable only with nightly rust...

But anyways this looks good to me in principle. Need to do a deeper review of the implementation but didn't look bad on my previous skim :)

File: bindgen/ir/comp.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsgf) [reply bindgen/ir/comp.rs:837]: Ah I parsed that ? as a question rather than an operator.

**Comment by pvdrz** [comment]
yeah we literally have the `experimental` flag on the CLI and a feature as well. I think other features are gated behind it already.

**Comment by jsgf** [comment]
Does `--experimental` do anything? I couldn't see anywhere that referenced it.

**Comment by jsgf** [comment]
@emilio @pvdrz Ping? Are there any outstanding concerns about this?

**Comment by jsgf** [comment]
Fixed the consistency of the `flex_mut_ref` vs `flex_ptr_mut` names (went with trailing `_mut`) and inlined the helpers since they should have effectively no runtime implementation (they're really typesystem operations).

**Comment by jsgf** [comment]
@emilio @pvdrz Ping?

**Review by emilio — APPROVED** [review]
Thanks, this looks fine. Sorry, last week has been short due to easter holidays :)

**Comment by emilio** [comment]
Please rebase and ping, happy to merge. Hopefully it's just a matter of regenerating test expectations. Sorry, this didn't get into the merge queue because it got configured right after I pressed the "Merge when ready" button :/

**Comment by jsgf** [comment]
Do you want me to squash the commits?

**Comment by emilio** [comment]
> Do you want me to squash the commits?

Nah, all good, thanks!

**Pull request merged**